# Assignment 5
### Do all four questions.

### Names: Erin Siedlecki, Shaveen Saadee, Anna Li, Emily Garman, Razan Habboub, Marissa Burton

**1.** Let's review some basic matrix multiplication. When you have an $M \times N$ matrix $A$ with $M$ rows and $N$ columns, 
$$
A= \left[ \begin{array}{cccc} a_{11} & a_{12} & ... & a_{1N} \\
a_{21} & a_{22} & ... & a_{2N} \\
\vdots & \vdots & ... & \vdots \\
a_{M1} & a_{M2} & ... & a_{MN} 
\end{array} \right],
$$
and you right-multiply it by a vector
$$
x = \left[ \begin{array}{c} x_1 \\ x_2 \\ \vdots \\ x_N 
\end{array} \right],
$$
you get
$$
Ax = \left[ \begin{array}{c} \sum_{i=1}^N a_{1i} x_i \\ \sum_{i=1}^N a_{2i} x_i \\ \vdots \\ \sum_{i=1}^N a_{Mi} x_i 
\end{array} \right].
$$
This is just "matrix row times column vector" element-by-element, stacking the results into a new vector.

For this to make sense, $N$ must be the same for the matrix and the vector, but $M$ can be different from $N$. 

Let's play with some NumPy to see this. First we'll define a matrix $A$:

In [19]:
import numpy as np

A = np.array([ [1,2,3],
              [4,5,6],
              [7,8,9]])
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

a. Multiply $A$ times each of the following vectors using the @ operator. Explain which part of the $A$ matrix gets selected and explain why, using the definition of matrix multiplication. 

In [20]:
e_1 = np.array([1,0,0])
e_2 = np.array([0,1,0])
e_3 = np.array([0,0,1])

In [21]:
A@e_1

array([1, 4, 7])

When multiplying A by [1,0,0], we multiply the first column of A by 1, the second column of A by 0, and the third column of A by 0 and add the results of each row. Because the second and third columns of A are multiplied by 0, their results are 0, so 0 will be added to each row in the first column of A, which just returns the first column of A. Thus, our final answer is just the first column of A, which is [1,4,7].

In [22]:
A@e_2

array([2, 5, 8])

Similar to the last multiplication, when multiplying A by [0,1, 0], we multiply the first column of A by 0, the second column of A by 1, and the third column of A by 0 and add the results of each row. Because the first and third columns of A are multiplied by 0, their results are 0, so 0 will be added to each row in the second column of A, which just returns the second column of A. Thus, our final answer is just the second column of A, which is [2,5,8].

In [23]:
A@e_3

array([3, 6, 9])

As we saw with the other two multiplications, when multiplying A by [0,0,1], we multiply the first column of A by 0, the second column of A by 0, and the third column of A by 1 and add the results of each row. Because the first and second columns of A are multiplied by 0, their results are 0, so 0 will be added to each row in the third column of A, which just returns the third column of A. Thus, our final answer is just the third column of A, which is [3,6,9].

b. Now multiply $A$ times $u = (1,1,1)$. Explain the logic of the result with the definition of matrix multiplication.

In [24]:
u = np.ones(3)

In [25]:
A@u

array([ 6., 15., 24.])

When multiplying A by [1,1,1], we multiply the first, second, and third columns of A by 1 and add the results of each row together. This gives us 1+2+3 in the first row, 4+5+6 in the second row, and 7+8+9 in the third row, which results in [6,15,24].

c. Whenever a matrix has 1's on the diagonal and zeros everywhere else, we call it an **identity matrix**. What happens when you multiple $A$ times $x$ below? What happens when you multiple an identity matrix times any vector? Explain your result with the definition of matrix multiplication.

In [26]:
A = np.array([ [1,0,0],
              [0,1,0],
              [0,0,1]])
x = np.array([-2,4,11])


In [27]:
A@x

array([-2,  4, 11])

When you multiply the identity matrix A by [-2,4,11], we multiply the first column of A by -2, the second column of A by 4, and the third column of A by 11, and add the results of each row. Because the first row of A has a 1 in the first column and 0s in the second and third columns, we get (-2x1) + 0 + 0, and the result of adding these together is -2. The second row of A has a 1 in the second column and 0s in the first and third columns, so we get 0 + (1x4) + 0, and the result of adding these together is 4. The third row of A has 1 in the third column and 0s in the first and second columns, so we get 0 + 0 + (1x11), and the result of adding these together is 11. Thus, our final result is [-2,4,11].

d. What if every row and column sum to 1, but the 1's are no longer on the diagonal? Multiple $A$ times $X$ below and explain the result. Create another matrix whose rows and columns sum to 1, but is not an identity matrix, and show how it permutes the values of $x$. 

In [28]:
A = np.array([ [0,0,1],
              [1,0,0],
              [0,1,0]])
x = np.array([-2,4,11])


In [29]:
A@x

array([11, -2,  4])

If every row and column sum to 1, but the 1s are not on the diagonal, the same multiplication principles apply and we see the result is the corresponding value of x that is multiplied by the column that has the 1 in it. In this case, we see that the first row of A has the 1 in the third column, which is multiplied by the third element in x, thus our result for that row (since the other elements are 0) is 11. For the second row, the 1 is in the first column of A, which is multipled by the first element of x, thus our result for the second row (since the other elements are 0) is -2. The last row of A has the 1 in the second column, which is multipled by the second element of x, thus our result for the third row (since the other elements are 0) is 4. Our final result is [11,-2,4].

In [30]:
A = np.array([[0,1,0],
              [0,0,1],
              [1,0,0]])
x = np.array([-2,4,11])

A@x

array([ 4, 11, -2])

Similar to what we saw before, since each row and column only have one 1 and the rest 0s, we see the result of our matrix multiplication is just the corresponding value of x that is multiplied by the column in the row that has the 1. For the first row of A, the second column has the 1, so the result is the second element of x, which is 4. For the second row of A, the third column has the 1, so the result is the third element of x, which is 11. For the third row of A, the first column has the 1, so the result is the first element of x, which is -2. Thus our final result is [4,11,-2].

e. The next matrix $A$ could be a Markov transition matrix: Its columns sum to 1, and each entry $a_{ij}$ can be interpreted as the proportion of observations who moved from state $j$ to state $i$. Multiply $A$ by each of the vectors $e_1$, $e_2$, and $e_3$, and explain your results.

In [31]:
rng = np.random.default_rng(100)
A = rng.random((3,3)) # Generate a random 3X3 matrix
sums = np.sum(A,axis=0) # Column sums
A = A/sums # Normalize the columns so they sum to 1
print(A)

[[0.50052958 0.24049286 0.18358131]
 [0.02574731 0.39251588 0.37907577]
 [0.47372311 0.36699127 0.43734292]]


In [32]:
A@e_1

array([0.50052958, 0.02574731, 0.47372311])

In [33]:
A@e_2

array([0.24049286, 0.39251588, 0.36699127])

In [34]:
A@e_3

array([0.18358131, 0.37907577, 0.43734292])

Similar to the results we saw in part a, when multiplying A by e_1 which is [1,0,0], we multiply the first column of A by 1 (since the 1 is in the first row of e_1), the second column of A by 0, and the third column of A by 0 and add the results of each row. Because the second and third columns of A are multiplied by 0, 0s will be added to each row in the first column of A, which just returns the first column of A. Thus, our final answer is just the first column of A, which is [0.50052958, 0.02574731, 0.47372311]. The same pattern applies for multiplying A and e_2. Since the 1 is in the second row of e_2 and the rest are 0s, our result is going to be the second column of A, which is [0.24049286, 0.39251588, 0.36699127]. Lastly, when multiplying A and e_3, since the 1 is in the third row of e_2 and the rest are 0s, our result is going to be the third column of A, which is [0.18358131, 0.37907577, 0.43734292].

f. For each of the vectors $e_1, e_2, e_3$, multiple $A$ times that vector 5 times. What answer do you get for each starting vector? Describe the behavior you observe.

In [35]:
for i, e in enumerate([e_1, e_2, e_3], start=1):
    v = e.copy()
    print(f"\nVector e_{i}:")
    for k in range(5):
        v = A @ v
        print(f"A^{k+1} e_{i} = {v}")


Vector e_1:
A^1 e_1 = [0.50052958 0.02574731 0.47372311]
A^2 e_1 = [0.34368862 0.20257047 0.45374091]
A^3 e_1 = [0.30404142 0.26036337 0.43559521]
A^4 e_1 = [0.29476439 0.27514859 0.43008701]
A^5 e_1 = [0.29266551 0.27862515 0.42870935]

Vector e_2:
A^1 e_2 = [0.24049286 0.39251588 0.36699127]
A^2 e_2 = [0.28214379 0.29937825 0.41847795]
A^3 e_2 = [0.29004438 0.28341002 0.42654561]
A^4 e_2 = [0.29163968 0.2804039  0.42795643]
A^5 e_2 = [0.29197422 0.27979983 0.42822595]

Vector e_3:
A^1 e_3 = [0.18358131 0.37907577 0.43734292]
A^2 e_3 = [0.26334088 0.31930609 0.41735304]
A^3 e_3 = [0.28521895 0.29032145 0.4244596 ]
A^4 e_3 = [0.29050361 0.28220175 0.42729464]
A^5 e_3 = [0.29171646 0.2802254  0.42805814]


Each vector grows quickly in magnitude, and after a few multiplications, all three vectors become almost proportional to the same vector, aligning with the matrix's principal eigenvector.

*2.* Let's consider a simple Markov transition matrix over two states:
$$
T = \left[ \begin{array}{cc} p_{1\leftarrow 1} &  p_{1\leftarrow 2} \\
p_{2 \leftarrow 1} & p_{2 \leftarrow 2} \end{array}\right] 
$$
The arrows help visualize the transition a bit: This is the same index notation as usual, $p_{ij}$, but writing it $p_{i \leftarrow j}$ emphasizes that it's the proportion of times that state $j$ transitions to state $i$. Below, $T$ is given by
$$
T = \left[ \begin{array}{cc} .25 & .5 \\
.75 & .5 \end{array}\right].
$$

- Start in state 1, at the initial condition $[1,0]$. Multiply that vector by $T$. Write out the result in terms of the formula and compute the result in a code chunk below. What is this object you're looking at, in terms of proportions and transitions?
- Multiple by $T$ again. What do you get? This isn't a column of $T$. Explain in words what it is. (Hint: A forecast of what in what period?)
- Keep multiplying the current vector of outcomes by $T$. When does it start to settle down without changing further?
- Do the above analysis again, starting from the initial condition $[0,1]$. Do you get a different result?
- The take-away is that, in the long run, these chains settle down into the long-run proportions, and the sensitivity on initial conditions vanishes. 


In [ ]:
T = np.array([[ 1/4, 1/2],
                 [ 3/4, 1/2 ]])

: 

3. Weather data

- Load the `cville_weather.csv` data. This includes data from Jan 4, 2024 to Feb 2, 2025. Are there any missing data issues?
- Based on the precipitation variable, `PRCP`, make a new variable called `rain` that takes the value 1 if `PRCP`>0 and 0 otherwise.
- Build a two-state Markov chain over the states 0 and 1 for the `rain` variable. 
- For your chain from c, how likely is it to rain if it was rainy yesterday? How likely is it to rain if it was clear yesterday?
- Starting from a clear day, forecast the distribution. How quickly does it converge to a fixed result? What if you start from a rainy day?
- Conditional on being rainy, plot a KDE of the `PRCP` variable.
- Describe one way of making your model better for forecasting and simulation the weather.

Congratulations, you now are a non-parametric meteorologist!

In [ ]:
import pandas as pd
import numpy as np

weather = pd.read_csv('./DS-5030/cville_weather.csv', low_memory=False)

: 

In [ ]:
weather.isna().sum()

STATION              0
NAME                 0
DATE                 0
DAPR               399
DAPR_ATTRIBUTES    399
MDPR               399
MDPR_ATTRIBUTES    399
PRCP                12
PRCP_ATTRIBUTES     12
SNOW               188
SNOW_ATTRIBUTES    188
SNWD               410
SNWD_ATTRIBUTES    410
dtype: int64

: 

Yes, there are missing data issues. There are a significant amount of NaNs present in the data.

In [ ]:
weather['rain'] = (weather['PRCP'] > 0).astype(int)

: 

In [ ]:
T = np.array([['P00', 'P01'],
['P10', 'P11']])
T

array([['P00', 'P01'],
       ['P10', 'P11']], dtype='<U3')

: 

In [ ]:
rain = weather['rain']

rains = np.zeros((2, 2))

for (a,b) in zip(rain[:-1], rain[1:]):
    rains[int(a), int(b)] += 1

T = rains / rains.sum(axis=1, keepdims=True)

: 

In [ ]:
T

array([[0.73160173, 0.26839827],
       [0.35195531, 0.64804469]])

: 

If it was rainy yesterday, the probability that it is raining today is 0.73160173. If it was clear yesterday, the probability that it will be rainy today is 0.35195531.

In [ ]:
# Starting from a clear day, forecast the distribution.

n = 100
p = np.array([1,0])
for t in range(n):
    print(p)
    p = p @ T

[1 0]
[0.73160173 0.26839827]
[0.62970529 0.37029471]
[0.59102067 0.40897933]
[0.57633419 0.42366581]
[0.57075852 0.42924148]
[0.56864174 0.43135826]
[0.56783811 0.43216189]
[0.56753302 0.43246698]
[0.56741719 0.43258281]
[0.56737321 0.43262679]
[0.56735652 0.43264348]
[0.56735018 0.43264982]
[0.56734777 0.43265223]
[0.56734686 0.43265314]
[0.56734651 0.43265349]
[0.56734638 0.43265362]
[0.56734633 0.43265367]
[0.56734631 0.43265369]
[0.56734631 0.43265369]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463

: 

In [ ]:
# Starting from a rainy day, forecast the distribution.
n = 100
p = np.array([0,1])
for t in range(n):
    print(p)
    p = p @ T

[0 1]
[0.35195531 0.64804469]
[0.48557388 0.51442612]
[0.5363017 0.4636983]
[0.55556033 0.44443967]
[0.5628718 0.4371282]
[0.56564757 0.43435243]
[0.56670139 0.43329861]
[0.56710146 0.43289854]
[0.56725335 0.43274665]
[0.56731101 0.43268899]
[0.5673329 0.4326671]
[0.56734122 0.43265878]
[0.56734437 0.43265563]
[0.56734557 0.43265443]
[0.56734602 0.43265398]
[0.5673462 0.4326538]
[0.56734626 0.43265374]
[0.56734629 0.43265371]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537]
[0.5673463 0.4326537

: 

These converge quickly (after about 20 iterations).

In [ ]:
import matplotlib.pyplot as plt

: 

In [ ]:
# Conditional on being rainy, plot a KDE of the `PRCP` variable.

rainy_days = weather[weather['rain'] == 0]
rainy_days['PRCP'].plot.kde()
plt.xlabel("Precipitation (PRCP)")
plt.show()

KeyError: 'rain'

: 

To make our model better for forecasting and simulating the weather, [insert more].


4. Taxicab trajectories: Using the pickled taxicab data, we want to complete the exercise from class.

- For the taxicab trajectory data, determine your state space and clean your sequences of cab rides.
- Compute the transition matrix for the taxicab data between neighborhoods in Manhattan. Plot it in a heat map. What are the most common routes?
- Explain why taxicabs are most likely order 1, and not 2 or more.
- Starting at Hell's Kitchen, create a sequence of forecasts of where the cab is likely to be in 2, 3, 5, and 10 trips
- Starting at any neighborhood, iterate your forecast until it is no longer changing very much. Where do cabs spend most of their time working in Manhattan?